src/Microsoft.OData.Client/Metadata/ClientTypeUtil.cs

```diff
@@ -439,70 +442,80 @@ internal static PropertyInfo[] GetKeyPropertiesOnType(Type type)
+                Type keyPropertyDeclaringType = null;
+                foreach (PropertyInfo key in keyProperties)
                 {
-                    throw c.Error.InvalidOperation(c.Strings.ClientType_KeysMustBeSimpleTypes(key.Name, typeName, key.PropertyType.FullName));
+                    if (keyPropertyDeclaringType == null)
+                    {
+                        keyPropertyDeclaringType = key.DeclaringType;
+                    }
+                    else if (keyPropertyDeclaringType != key.DeclaringType)
+                    {
+                        throw c.Error.InvalidOperation(c.Strings.ClientType_KeysOnDifferentDeclaredType(typeName));
+                    }
+
+                    if (!PrimitiveType.IsKnownType(key.PropertyType) && !(key.PropertyType.GetGenericTypeDefinition() == typeof(System.Nullable<>) && key.PropertyType.GetGenericArguments().First().IsEnum()))
+                    {
+                        throw c.Error.InvalidOperation(c.Strings.ClientType_KeysMustBeSimpleTypes(key.Name, typeName, key.PropertyType.FullName));
+                    }
                 }
-            }
 
-            if (dataServiceKeyAttribute != null)
-            {
-                if (newKeyKind == KeyKind.AttributedKey && keyProperties.Count != dataServiceKeyAttribute?.KeyNames.Count)
+                if (dataServiceKeyAttribute != null)
                 {
-                    var m = (from string a in dataServiceKeyAttribute.KeyNames
-                             where (from b in properties
-                                    where b.Name == a
-                                    select b).FirstOrDefault() == null
-                             select a).First<string>();
-                    throw c.Error.InvalidOperation(c.Strings.ClientType_MissingProperty(typeName, m));
+                    if (newKeyKind == KeyKind.AttributedKey && keyProperties.Count != dataServiceKeyAttribute?.KeyNames.Count)
+                    {
+                        var m = (from string a in dataServiceKeyAttribute.KeyNames
+                                 where (from b in properties
+                                        where b.Name == a
+                                        select b).FirstOrDefault() == null
+                                 select a).First<string>();
+                        throw c.Error.InvalidOperation(c.Strings.ClientType_MissingProperty(typeName, m));
+                    }
                 }
+
+                typeInfo.KeyProperties = keyProperties.Count > 0 ? keyProperties.ToArray() : (isEntity ? ClientTypeUtil.EmptyPropertyInfoArray : null);
             }
 
-            return keyProperties.Count > 0 ? keyProperties.ToArray() : (isEntity ? ClientTypeUtil.EmptyPropertyInfoArray : null);
+            hasProperties = typeInfo.HasProperties.Value;

```

Comment:
hasProperties is a local variable or the instance variable?
What's the meaning here to set its value?

Reply:
out param